File: 0134-gas-station/0134-gas-station.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        
        int length = gas.length;
        int i = 0;
        while (i < length) {
            if (gas[i] < cost[i]) {
                i ++;
            }

            int remain = 0;
            boolean possible = true;

            int steps = 0;
            int j = i;

            while (steps < length) {
                remain += gas[j] - cost[j];

                if (remain < 0) {
                    possible = false;
                    break;
                }

                j = (j + 1) % length;
                steps++;
            }

            if (possible) return i;

            i += steps;
        }

        return -1;
    }
}

File: 0134-gas-station/0134-gas-station.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int len = gas.size();
        int i = 0;
        while (i < gas.size()) {
            
            int currIdx = i;
            int remain = 0;

            do {
                
                remain = remain + gas[currIdx] - cost[currIdx];

                if (remain < 0) break;

                currIdx = (currIdx + 1) % len;
            } while (currIdx != i);


            if (remain >= 0) return i;
            if (currIdx < i) break;

            i = currIdx + 1;
        }

        return -1;
    }


};